permission,dns: add permission for dns

theanarkh opened the pull request:
Refs: https://github.com/nodejs/node/pull/44004

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

marco-ippolito commented:
I assume this also impacts the http module

anonrig commented on src/cares_wrap.h:
```diff
@@ -233,7 +233,6 @@ class QueryWrap final : public AsyncWrap {
       : AsyncWrap(channel->env(), req_wrap_obj, AsyncWrap::PROVIDER_QUERYWRAP),
         channel_(channel),
         trace_name_(Traits::name) {}
-
```
Can you remove these irrelevant changes?

theanarkh replied:
Done!

theanarkh commented:
> I assume this also impacts the http module

Yes. Or we can skip the permission check when using dns module in Node.js core.

marco-ippolito commented:
> > I assume this also impacts the http module
> 
> Yes. Or we can skip the permission check when using dns module in Node.js core. 

Its very important to document it as I think http is unusable without dns being granted

RafaelGSS commented on doc/api/cli.md:
```diff
@@ -357,6 +357,48 @@ Error: Access to this API has been restricted
+    at node:internal/main/run_main_module:30:49 {
+  code: 'ERR_ACCESS_DENIED',
+  permission: 'NetDNS',
+  resource: 'lookup'
```
The resource should be the lookup address, in this case: `localhost`.

RafaelGSS commented on doc/api/cli.md:
````diff
@@ -357,6 +357,48 @@ Error: Access to this API has been restricted
+```
+
+```console
+$ node --experimental-permission --allow-fs-read=\* index.js
````
```suggestion
$ node --experimental-permission --allow-fs-read=./index.js index.js
```

RafaelGSS commented on doc/node.1:
```diff
@@ -94,6 +94,9 @@ Allow execution of WASI when using the permission model.
 Allow creating worker threads when using the permission model.
 .
+.It Fl -allow-net-dns
+Allow execution of DNS when using the permission model.
```
```suggestion
Allow execution of DNS Lookup when using the permission model.
```

RafaelGSS commented on doc/api/permissions.md:
```diff
@@ -575,6 +576,7 @@ There are constraints you need to know before using this system:
   * Inspector protocol
   * File system access
   * WASI
+  * DNS
```
```suggestion
  * DNS Lookup
```

RafaelGSS commented on lib/internal/process/pre_execution.js:
```diff
@@ -530,6 +530,7 @@ function initializePermission() {
     const warnFlags = [
       '--allow-addons',
       '--allow-child-process',
+      '--allow-net-dns',
```
I don't think we should emit a warning when `--allow-net-dns` is used. It won't open breaches to bypass permission model restrictions as the other CLI args.

RafaelGSS commented on doc/api/cli.md:
````diff
@@ -357,6 +357,48 @@ Error: Access to this API has been restricted
 }
 ```
 
+### `--allow-net-dns`
````
I think instead of being a _true_/_false_ option, we should accept a list of DNS allowed to lookup.

RafaelGSS reviewed (changes requested):
Hey, thanks for working on that.

Would you mind providing some context on why you believe this is useful for end users? In other words, what kind of security would this provide?

Requesting changes just to make sure we are aligned this is something we'd like to see on the Permission Model

ShogunPanda commented:
> > > I assume this also impacts the http module
> > 
> > 
> > Yes. Or we can skip the permission check when using dns module in Node.js core.
> 
> Its very important to document it as I think http is unusable without dns being granted

I would say it's 90% unusable. AFAIK `node:http` does not use DNS internally but `node:net` does. So unless a person only uses IP based communication (that would be... interesting :D) I think you are 110% right.

Anyway it's not only `http`. Even the `net` or `tls` would make way less sense without DNS.
Do we want to go for `--allow-net` instead?

theanarkh commented:
Thank you for reviewing the code. I tried to add permission verification to the network(TCP / UDP / DNS) module. The user can set the access to only certain domain names or IP addresses, for example, to avoid SSRF attacks.

ShogunPanda commented:
@theanarkh I think I'm losing some context. You said you blocked access but I don't see in the code. Are you saying that you blocked at DNS level, isn't it? If that's the case, I was instead talking to prevent `net.connect` or `net.createServer` as well.

marco-ippolito commented:
I see a lot of value in this PR, as part of adding permission model to the network.
My only concern is documenting behavior and edge cases when interacting with net module

ShogunPanda commented:
@marco-ippolito I totally agree. If we agree this is only the first PR which will be shortly followed by others for complete coverage of `net` module, then I'm fine with it.

But I get to choose I'd prefer a single and comprehensive PR that covers both `dns` and `net`  module in a single CLI flag.

Nonetheless, for clarity I'm not blocking this as it is an amazing piece of work ❤️

theanarkh commented:
Thanks for all the suggestions, I convert this PR to draft and continue to work on it.

RafaelGSS commented:
> I probably see it more as part of --allow-net rather then dns by itself

I see that too, it's probably more useful to block all network communication instead of specific DNS operations.

ShogunPanda commented:
Some point I noted here.

1. (Blocking) Do not use `$IP/$PORT` syntax anywhere. People are expecting the `:` separator, as `/` is used for netmask (see below)
2. (Blocking) When going through IP, I think we should support netmask, both in bitmask (12.34.56.78/255.255.255.0) and numeric (12.34.56.78/24) mode.
3. (Non blocking) How does it work when people do something like `--allow-net-udp udp.host.com:1234`? Technically speaking the connection for the DNS resolving was not explicitly allowed.
4. (Non blocking) Do you plan to work on a follow up complementary PR for TCP?

RafaelGSS commented on doc/api/cli.md:
```diff
@@ -357,6 +357,107 @@ Error: Access to this API has been restricted
+
+> Stability: 1.1 - Active development
+
+When using the [Permission Model][], the process will not be able to bind or connect
```
Why is this explicitly to `bind` and `connect`? Can't we deny UDP communication completely (using nodejs APIs of course)?

If so, I think we should reword that to: `the process will not be able to create, receive, and, send UDP packages.`

RafaelGSS commented on doc/node.1:
```diff
@@ -94,6 +94,15 @@ Allow execution of WASI when using the permission model.
 Allow creating worker threads when using the permission model.
 .
+.It Fl -allow-net-udp
+Allow UDP bind or connect when using the permission model.
```
```suggestion
Allow UDP bind and connect when using the permission model.
```

RafaelGSS reviewed (commented):
I will need some time to test it properly, but I'm wondering if we should have `udp.in` and `udp.out` at this point. I think this kind of granularity can be very useful once we have a stable  API for `--allow-net`. What do you think about removing `--allow-net-udp-in` and `--allow-net-udp-out` in favor of `--allow-net-udp`?

marco-ippolito commented:
> I will need some time to test it properly, but I'm wondering if we should have `udp.in` and `udp.out` at this point. I think this kind of granularity can be very useful once we have a stable API for `--allow-net`. What do you think about removing `--allow-net-udp-in` and `--allow-net-udp-out` in favor of `--allow-net-udp`?

I do agree, the general use case is that if I dont want my application to access the network, I'd rather have `--allow-net` and then if necessary make it more granular, than having to `--allow-net-upd-in`, `--allow-net-udp-out`, `--allow-net-tcp-in`, `--allow-net-tcp-out`

theanarkh commented:
> First of all, thanks for this. This is huge and very welcome! ❤️ Some point I noted here.
> 
> 1. (Blocking) Do not use `$IP/$PORT` syntax anywhere. People are expecting the `:` separator, as `/` is used for netmask (see below)
> 2. (Blocking) When going through IP, I think we should support netmask, both in bitmask (12.34.56.78/255.255.255.0) and numeric (12.34.56.78/24) mode.
> 3. (Non blocking) How does it work when people do something like `--allow-net-udp udp.host.com:1234`? Technically speaking the connection for the DNS resolving was not explicitly allowed.
> 4. (Non blocking) Do you plan to work on a follow up complementary PR for TCP?

1.2 Thanks, i will use ":" to separate `$IP/$PORT` and support netmask.
3. `--allow-net-udp udp.host.com:1234` will only allow `udp.host.com:1234` and deny all the other hosts and IPs(before DNS lookup).
4. Yes. If this PR is merged i will open another PR for TCP.

theanarkh commented:
> I will need some time to test it properly, but I'm wondering if we should have `udp.in` and `udp.out` at this point. I think this kind of granularity can be very useful once we have a stable API for `--allow-net`. What do you think about removing `--allow-net-udp-in` and `--allow-net-udp-out` in favor of `--allow-net-udp`?

Thanks. I am fine to remove `--allow-net-udp-in` and `--allow-net-udp-out`. I will update the code later.

ShogunPanda commented on doc/api/cli.md:
```diff
@@ -357,6 +357,57 @@ Error: Access to this API has been restricted
+      throw er; // Unhandled 'error' event
+      ^
+
+Error [ERR_ACCESS_DENIED]: Access to this API has been restricted. Permission: bind to 127.0.0.1/9297
```
I think it should be `127.0.0.1:9297` here.

theanarkh replied:
If separated by `:`, it is difficult to parse out the IP and port when using IPv6 (such as ::1:80). And we need to something like this.
```js
if (isIPV6(ip)) {
    permission.has(`[ipv6]:port`)
} else {
   permission.has(`ipv4:port`)
}
```

ShogunPanda replied:
Ok, I would say: let's use `/` internally but make sure all outfacing error messages use `:`. Is that fine?

theanarkh replied:
Yeah ! It is fine.

ShogunPanda commented on lib/dgram.js:
```diff
@@ -372,7 +371,18 @@ Socket.prototype.bind = function(port_, address_ /* , callback */) {
-  });
+  };
+  if (permission.isEnabled()) {
+    const resource = `${address}/${port || '*'}`;
```
Same here, use `:`.

ShogunPanda commented on lib/dgram.js:
```diff
@@ -413,13 +423,27 @@ function _connect(port, address, callback) {
 
   const afterDns = (ex, ip) => {
+    if (!ex && !address && permission.isEnabled()) {
+      const resource = `${ip}/${port}`;
```
Same here, use `:`.

ShogunPanda commented on lib/dgram.js:
```diff
@@ -413,13 +423,27 @@ function _connect(port, address, callback) {
   };
-
+  if (address && permission.isEnabled()) {
+    const resource = `${address}/${port}`;
```
Same here, use `:`.

ShogunPanda commented on lib/dgram.js:
```diff
@@ -663,6 +691,13 @@ Socket.prototype.send = function(buffer,
   const afterDns = (ex, ip) => {
+    // If we have not checked before dns, check it now
+    if (!ex && !connected && !address && permission.isEnabled()) {
+      const resource = `${ip}/${port}`;
```
Same here, use `:`.

ShogunPanda commented on lib/dgram.js:
```diff
@@ -671,6 +706,16 @@ Socket.prototype.send = function(buffer,
   if (!connected) {
+    // If address is not empty, check it
+    if (address && permission.isEnabled()) {
+      const resource = `${address}/${port}`;
```
Same here, use `:`.

ShogunPanda reviewed (approved):
Few nits but LGTM.

RafaelGSS commented on doc/api/cli.md:
```diff
@@ -357,6 +357,57 @@ Error: Access to this API has been restricted
+    at afterDns (node:dgram:337:12)
+    at node:dgram:379:9
+    at process.processTicksAndRejections (node:internal/process/task_queues:77:11) {
+  code: 'ERR_ACCESS_DENIED'
```
Instead of adding `"Permission: bind to 127.0.0.1/9297"`, we should add a `resource` field and a `permission` field to the error object. See how --allow-fs-read works

RafaelGSS commented on src/permission/net_permission.cc:
```diff
@@ -0,0 +1,237 @@
+                   &netmask,
+                   result->netmask,
+                   ip_version)) {  // 127.0.0.1/255.255.255.0
+      std::cerr << "invalid netmask: " << result->netmask << std::endl;
```
It must be Node.js debug log

RafaelGSS reviewed (commented):
I believe the checks you added to `dgram` are way more assertive if placed into the native side (`udp_wrap`). I think we must have to move it here.

RafaelGSS commented:
Good work!

theanarkh commented:
> I believe the checks you added to `dgram` are way more assertive if placed into the native side (`udp_wrap`). I think we must have to move it here.

If add permission check in native side, i think we should pass a flags to tell native side if it needs to check the permission. Because the native side should not check the IP if the IP is from `dns.lookup`(such as --allow-net-udp=localhost).

In addition. If we use UDP in cluster worker(use `--allow-net-udp=xxx` flag), Node.js will call `bind` in main process(no permission flag) which will bypass the permission check. 

So i think we also need to check the permission in JS layer.

RafaelGSS commented:
> > I believe the checks you added to `dgram` are way more assertive if placed into the native side (`udp_wrap`). I think we must have to move it here.
> 
> If add permission check in native side, i think we should pass a flags to tell native side if it needs to check the permission. Because the native side should not check the IP if the IP is from `dns.lookup`(such as --allow-net-udp=localhost).
> 
> In addition. If we use UDP in cluster worker(use `--allow-net-udp=xxx` flag), Node.js will call `bind` in main process(no permission flag) which will bypass the permission check.
> 
> So i think we also need to check the permission in JS layer.

Hey, I'll review it carefully over the week. Sorry for the delay. Since it's a security feature, I want to make sure we have covered all caveats as it would expose security vulnerabilities if landed.

RafaelGSS reviewed (changes requested):
We need to adjust a few things before landing it. Assuming a regular udpv4 socket server:

```
const server = udp.createSocket('udp4');
server.on('error', () => {}
server.bind(() => {})
```

1. The error object should contain the `resource` property. In this case, _resource_ will be the address + port. 
2. The error object should contain the `permission` property. In this case, `NetUDP`.
3. How would we handle IPV6 addresses when `--allow-net-udp=IPV4/22`? Will Octal IP addresses be covered? We must have tests for all those assertions regardless of whether we are supporting this first iteration or not.
4. You need to rebase on main, since the `ERR_ACCESS_DENIED` error changed a bit (https://github.com/nodejs/node/commit/b9289a6e29e54beeaa3f781bde1195e48df0da75#diff-670bf55805b781d9a3579f8bca9104c04d94af87cc33220149fd7d37b095ca1cR1115).
5. We should not use `std::cerr` (iostream) for error messages. Instead, create proper error objects.
6. What if `fd` is passed to `_createSocketHandle`? I don't think we could support it. But, we must document. 
7. Last but not least, can't we move all those checks to the C++ side? I think it's less susceptible to changes that would bypass the UDP Check. For instance, if someone writes a new JS API to create UDP sockets, they might forget to include `permission.has` check leading to an exposure. If we use the `udp_wrap.cc` it's more likely to be reused in any new JS API.


Overall, looks really good.

theanarkh commented:
1,2,4: I have rebased the main branch into my branch.

3: What does the `Will Octal IP addresses be covered?` mean ? Binding to IPV6 address will throw an error when only allow IPV4 address. For example:
```js
const dgram = require('dgram')
const server = dgram.createSocket('udp4');
server.on('error', console.log)
server.bind(8888, "::1",() => {})
```
`node --experimental-permission --allow-fs-read=./app.js --allow-net-udp=127.0.0.1:8888 app.js`. The output is as follows:
```
Error [ERR_ACCESS_DENIED]: bind to ::1/8888
    at node:dgram:379:18
    at process.processTicksAndRejections (node:internal/process/task_queues:77:11) {
  code: 'ERR_ACCESS_DENIED',
  permission: '::1/8888',
  resource: 'NetUDP'
}
```

5. I have updated the code to check the options in `EnvironmentOptions::CheckOptions`.

6. Yes. We do not support this, i will update the document.

7. I can add a check in C++ layer, but i think we can't just check in the C++ layer. For example:
```js
const dgram = require('dgram')
const server = dgram.createSocket('udp4');
server.on('error', console.log)
server.bind(8888, "localhost",() => {})
```
`node --experimental-permission --allow-fs-read=./app.js --allow-net-udp=localhost:8888 app.js`

If we only check in C++ layer, the code above will throw an error because we do not allow bind to 127.0.0.1(the result of dns.lookup).

Thanks for reviewing. I will continue to improve the code and add more tests.

anonrig commented on src/permission/net_permission.cc:
```diff
@@ -0,0 +1,172 @@
+    // Allow multiple item in one flag, like --allow-net-udp=127.0.0.1,localhost
+    const std::vector<std::string_view> addresses = SplitString(res, ","sv);
+    for (const auto& address : addresses) {
+      switch (scope) {
```
We don't need this switch. We can just keep it as a simple if statement to simplify the code here.

anonrig commented on src/permission/net_permission.cc:
```diff
@@ -0,0 +1,172 @@
+}
+
+void NetPermission::GrantAccess(PermissionScope scope,
+                                const std::string_view& param) {
```
No need to pass the reference here. By default it doesn't copy. String view is basically 2 pointers.

```suggestion
                                const std::string_view param) {
```

anonrig commented on src/permission/net_permission.cc:
```diff
@@ -0,0 +1,172 @@
+
+bool NetPermission::is_granted(Environment* env,
+                               PermissionScope perm,
+                               const std::string_view& param = "") const {
```
```suggestion
                               const std::string_view param = "") const {
```

anonrig commented on src/permission/net_permission.cc:
```diff
@@ -0,0 +1,172 @@
+
+bool NetPermission::CheckoutResource(
+    const std::vector<std::unique_ptr<NetNode>>& list,
+    const std::string_view& param) const {
```
```suggestion
    const std::string_view param) const {
```

anonrig commented on src/permission/net_permission.cc:
```diff
@@ -0,0 +1,172 @@
+  const std::vector<std::string_view> address_port = SplitString(param, "/"sv);
+  int size = address_port.size();
+  CHECK(size == 1 || size == 2);
+  std::string address;
+  std::string port;
```
This makes a copy. You can keep it as `std::string_view` until you explicitly need to copy.

anonrig commented on src/permission/net_permission.cc:
```diff
@@ -0,0 +1,172 @@
+  return false;
+}
+
+void NetPermission::Print() const {
+  if (UNLIKELY(per_process::enabled_debug_list.enabled(
+          DebugCategory::PERMISSION_MODEL))) {
+    std::cout << "net-udp-net: " << std::endl;
+    std::cout << "  deny_all_udp_: " << deny_all_udp_ << std::endl;
+    std::cout << "  allow_all_udp_: " << allow_all_udp_ << std::endl;
+    JSONWriter writer(std::cout, false);
+    writer.json_arraystart("Nodes");
+    for (uint32_t i = 0; i < granted_udp_.size(); i++) {
+      auto node = granted_udp_[i].get();
+      writer.json_start();
+      writer.json_keyvalue("address", node->address);
+      writer.json_keyvalue("port", node->port);
+      if (node->net_mask_node) {
+        writer.json_keyvalue("netmask",
+                             node->net_mask_node->netmask.to_string());
+        writer.json_keyvalue("network",
+                             node->net_mask_node->network.to_string());
+      }
+      writer.json_end();
+    }
+    writer.json_arrayend();
+    std::cout << std::endl;
+  }
+}
```
```suggestion
```

theanarkh replied:
This function only use when `per_process::enabled_debug_list.enabled(DebugCategory::PERMISSION_MODEL))` is true.

anonrig commented on src/permission/net_permission.h:
```diff
@@ -0,0 +1,60 @@
+ private:
+  void GrantAccess(PermissionScope scope, const std::string_view& param);
+  bool CheckoutResource(const std::vector<std::unique_ptr<NetNode>>& list,
+                        const std::string_view& param = "") const;
```
```suggestion
                        const std::string_view param = "") const;
```

theanarkh commented on lib/dgram.js:
```diff
@@ -372,7 +371,22 @@ Socket.prototype.bind = function(port_, address_ /* , callback */) {
+    }
+  }
+  // Resolve address first
+  state.handle.lookup(address, afterDns);
```
If the address which is allowed is a domain, we should allow bind to all IPs this domain ponit to, so I think we need check before DNS lookup.

theanarkh commented:
@anonrig Thanks for your review. I have updated the code and added some replies.

theanarkh commented on src/permission/net_permission.cc:
```diff
@@ -0,0 +1,172 @@
+                          const std::vector<std::string>& allow,
+                          PermissionScope scope) {
+  // For Debug
+  auto cleanup = OnScopeLeave([&]() { Print(); });
```
This function only use when per_process::enabled_debug_list.enabled(DebugCategory::PERMISSION_MODEL)) is true.

RafaelGSS commented on src/permission/net_permission.cc:
```diff
@@ -0,0 +1,172 @@
+
+namespace permission {
+
+bool ConvertIPToBitset(std::bitset<IPV6_BIT_LEN>* bitset,
```
We should have a cpp test for this

RafaelGSS commented on src/udp_wrap.cc:
```diff
@@ -301,15 +301,26 @@ void UDPWrap::DoBind(const FunctionCallbackInfo<Value>& args, int family) {
   if (!args[1]->Uint32Value(ctx).To(&port) ||
       !args[2]->Uint32Value(ctx).To(&flags))
     return;
+  bool checked = true;
+  if (args.Length() > 3 && args[3]->IsBoolean()) {
+    checked = args[3].As<Boolean>()->Value();
+  }
```
Why is this needed?

RafaelGSS commented on src/util.cc:
```diff
@@ -861,4 +861,59 @@ v8::Maybe<int> GetValidFileMode(Environment* env,
   return v8::Just(mode);
 }
 
+bool IsDigit(const std::string_view str) {
```
All these functions needs a cpp test

theanarkh commented:
> > If we only check in C++ layer, the code above will throw an error because we do not allow bind to 127.0.0.1(the result of dns.lookup).
> 
> Yeah, that's expected. In the future, we can check if `--allow-net-udp=localhost` is passed, we add `127.0.0.1` to the allowed list too

But we can not know the IP of other domains. For example, when `--allow-net-udp=nodejs.org` is passed(the IPs of  `nodejs.org` is `1.1.1.1` and `2.2.2.2`), C++ layer(udp_wrap.cc) will  throw an error when check the `1.1.1.1` or `2.2.2.2` which is not allowed.

RafaelGSS commented:
> > > If we only check in C++ layer, the code above will throw an error because we do not allow bind to 127.0.0.1(the result of dns.lookup).
> > 
> > 
> > Yeah, that's expected. In the future, we can check if `--allow-net-udp=localhost` is passed, we add `127.0.0.1` to the allowed list too
> 
> But we can not know the IP of other domains. For example, when `--allow-net-udp=nodejs.org` is passed(the IPs of `nodejs.org` is `1.1.1.1` and `2.2.2.2`), C++ layer(udp_wrap.cc) will throw an error when check the `1.1.1.1` or `2.2.2.2` which is not allowed.

Why we simply don't resolve the domain in the C++ layer? Something like:

* nodejs startup with `--allow-net-udp=nodejs.org`
* NetPermission::Apply resolves nodejs.org to 1.1.1.1 and 2.2.2.2
* Store 1.1.1.1 & 2.2.2.2 & nodejs.org to the allowed list

---

Also, what happens if I pass:

* --allow-net-udp=nodejs.org
* nodejs.org uses 1.1.1.1 IP Address
* After a while nodejs.org changes the IP address (new cloud provider/machine) to 4.4.4.4
* Permission Model will compare 4.4.4.4 to 1.1.1.1 and will incorrectly deny access
